File: 백준 문제 풀이 모음/python/class 4/14938.py
import sys, heapq
input = sys.stdin.readline
INF = sys.maxsize

n, m, r = map(int, input().split())
item = list(map(int, input().split()))
graph = [[] for i in range(n+1)]

for i in range(r) :
    a, b, c = map(int, input().split())
    graph[a].append((b, c))
    graph[b].append((a, c))

def dijkstra(start) :
    distance = [INF] * (n+1)
    distance[start] = 0
    heap = [(start, 0)]

    while heap :
        #print('heap: ', heap)
        now_node, now_cost = heapq.heappop(heap)
        for next_node, next_cost in graph[now_node] :
            #print('next node: ' + str(next_node) + ' next_cost: ', str(next_cost))
            next_cost = next_cost + now_cost
            if distance[next_node] > next_cost :
                distance[next_node] = next_cost
                #print('distance: ', distance)
                heapq.heappush(heap, (next_node, next_cost))
    return(distance)

cost_list = []
for i in range(1, n+1) :
    arr = dijkstra(i)
    #print('arr: ', arr)
    cost = 0
    for j in range(1, len(arr)) :
        if arr[j] <= m :
            cost += item[j-1] 
    
    #print('result cost: ' + str(cost))
    cost_list.append(cost)

print(max(cost_list))

File: 백준 문제 풀이 모음/kotlin/class 4/14938.kt
import java.util.PriorityQueue

private const val INF = Integer.MAX_VALUE / 2
private lateinit var list : MutableList<MutableList<Node>>
private lateinit var item_list : List<Int>
private class Node (var NodeNum : Int, var weight : Int) : Comparable<Node> {
    override fun compareTo(other: Node): Int {
        return weight - other.weight
    }
}

fun main() = with(System.`in`.bufferedReader()){
    val input = readLine().split(' ').toList()
    val n = input[0].toInt()
    val m = input[1].toInt()
    val r = input[2].toInt()


    list = MutableList<MutableList<Node>> (n+1) { MutableList<Node>(0) { Node(0,0)} }

    item_list = readLine().split(' ').toList().map { it.toInt() }
    repeat(r) {
        val input = readLine().split(' ').toList()
        list[input[0].toInt()].add(Node(input[1].toInt(), input[2].toInt()))
        list[input[1].toInt()].add(Node(input[0].toInt(), input[2].toInt()))
    }

    var result = - 1
    for (i in 1 .. n) {
        result = Math.max(result, dikjstra(i, n, m))
    }

    println(result)

}

fun dikjstra(start: Int, n: Int, m :Int) : Int {
    val visited = MutableList<Boolean>(n+1) { false }
    val dist = MutableList<Int>(n+1) { INF }

    dist[start] = 0
    val pq = PriorityQueue<Node>()
    pq.offer(Node(start, 0))

    while (pq.isNotEmpty()) {
        val node = pq.poll()

        if (!visited[node.NodeNum]) {
            visited[node.NodeNum] = true

            list[node.NodeNum].forEach {
                if (!visited[it.NodeNum] && dist[it.NodeNum] > dist[node.NodeNum] + it.weight) {
                    dist[it.NodeNum] = dist[node.NodeNum] + it.weight
                    pq.offer(Node(it.NodeNum, dist[it.NodeNum]))
                }
            }
        }
    }

    var cost = 0
    for (i in 1 .. n) {
        if (dist[i] <= m) {
            cost += item_list[i-1]
        }
    }
//    println(dist)
    return cost
}
